org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/BuildPathManager.java

```diff
@@ -330,10 +331,14 @@ private void configureAttachedSourcesAndJavadoc(IMavenProjectFacade facade, Prop
 
         ArtifactKey aKey = desc.getArtifactKey();
         if(aKey != null) { // maybe we should try to find artifactKey little harder here?
-          boolean downloadSources = desc.getSourceAttachmentPath() == null && srcPath == null
-              && mavenConfiguration.isDownloadSources();
-          boolean downloadJavaDoc = desc.getJavadocUrl() == null && javaDocUrl == null
-              && mavenConfiguration.isDownloadJavaDoc();
+          // We should update a sources/javadoc jar for a snapshot in case they're already downloaded.
+          File plainFile = desc.getPath() != null ? desc.getPath().toFile() : null;
+          File srcFile = srcPath != null ? srcPath.toFile() : null;
+          boolean downloadSources = (srcPath == null && mavenConfiguration.isDownloadSources())
+              || (plainFile != null && plainFile.canRead() && srcFile != null && srcFile.canRead());
+          File javaDocFile = javaDocUrl != null ? getAttachedArtifactFile(aKey, CLASSIFIER_JAVADOC) : null;
+          boolean downloadJavaDoc = (javaDocUrl == null && mavenConfiguration.isDownloadJavaDoc())
+              || (plainFile != null && plainFile.canRead() && javaDocFile != null && javaDocFile.canRead());
```

Comment:
Isn't this condition going to re-download sources even f the plainFile was not modified?

Reply:
Yes. But it doesn't mean that a snapshot jar will be actually re-downloaded.
I have removed the check for "outdated" snapshot from this method because it was a  duplicated one. So now it only enables the "download" and `BuildPathManager.isOutdatedSnapshot()` decides if we really need to re-download a snapshot jar.

Reply:
OK, I guess that the underlying Maven API do all the necessary optimization. Let's wait for CI feedback and we'll merge.